spring-batch-core/src/main/java/org/springframework/batch/core/launch/support/SimpleJobOperator.java

```diff
@@ -391,6 +392,45 @@ public JobExecution abandon(JobExecution jobExecution) throws JobExecutionAlread
 		return jobExecution;
 	}
 
+	@Override
+	public JobExecution recover(JobExecution jobExecution) {
+		Assert.notNull(jobExecution, "JobExecution must not be null");
+
+		if (jobExecution.getExecutionContext().containsKey("recovered")) {
+			if (logger.isInfoEnabled()) {
+				logger.info("already recovered job execution: " + jobExecution);
+			}
+			throw new UnexpectedJobExecutionException("JobExecution is already recovered");
+		}
+
+		BatchStatus jobStatus = jobExecution.getStatus();
+		if (jobStatus == BatchStatus.COMPLETED || jobStatus == BatchStatus.ABANDONED) {
```

Comment:
`BatchStatus.UNKNOWN` is also non recoverable, see https://github.com/spring-projects/spring-batch/blob/d95397faf023ee3293ee10b41977231734a0f5d1/spring-batch-core/src/main/java/org/springframework/batch/core/step/AbstractStep.java#L291. I will add that on merge.